h2o-py/tests/testdir_algos/stackedensemble/pyunit_stackedensemble_max_runtime_secs.py

```diff
@@ -0,0 +1,88 @@
+#!/usr/bin/env python
+# -*- encoding: utf-8 -*-
+from __future__ import print_function
+
+
+import h2o
+import sys
+
+sys.path.insert(1, "../../../")  # allow us to run this standalone
+
+from h2o.grid import H2OGridSearch
+from h2o.estimators.gbm import H2OGradientBoostingEstimator
+from h2o.estimators.stackedensemble import H2OStackedEnsembleEstimator
+from tests import pyunit_utils as pu
+
+seed = 1
+
+
+def prepare_data():
+    fr = h2o.import_file(path=pu.locate("smalldata/junit/weather.csv"))
+    target = "RainTomorrow"
+    fr[target] = fr[target].asfactor()
+    ds = pu.ns(x=fr.columns, y=target, train=fr)
+
+    return ds
+
+
+def test_stackedensemble_propagates_the_max_runtime_secs():
+    max_runtime_secs = 5
+    hyper_parameters = dict()
+    hyper_parameters["ntrees"] = [1, 3, 5]
+    params = dict(
+        fold_assignment="modulo",
+        nfolds=3,
+        keep_cross_validation_predictions=True
+    )
+
+    data = prepare_data()
+
+    gs1 = H2OGridSearch(H2OGradientBoostingEstimator(**params), hyper_params=hyper_parameters)
+    gs1.train(data.x, data.y, data.train, validation_frame=data.train)
+
+    se = H2OStackedEnsembleEstimator(base_models=[gs1], max_runtime_secs=max_runtime_secs)
+    se.train(data.x, data.y, data.train)
+    metalearner = h2o.get_model(se.metalearner()["name"])
+
+    # metalearner has the set max_runtine_secs
+    assert metalearner.actual_params['max_runtime_secs'] <= max_runtime_secs
+    assert metalearner.actual_params['max_runtime_secs'] > 0
+
+    # stack ensemble has the set max_runtime_secs
+    assert se.max_runtime_secs == max_runtime_secs
+
+
+def test_stackedensemble_respects_the_max_runtime_secs():
+    max_runtime_secs = 1
+    hyper_parameters = dict()
+    hyper_parameters["ntrees"] = [1, 2, 3, 4, 5]
+    params = dict(
+        fold_assignment="modulo",
+        nfolds=3
+    )
+
+    data = prepare_data()
+
+    gs1 = H2OGridSearch(H2OGradientBoostingEstimator(**params), hyper_params=hyper_parameters)
+    gs1.train(data.x, data.y, data.train, validation_frame=data.train)
+
+    big_blending_frame = data.train
+    for i in range(15):
+        big_blending_frame = big_blending_frame.rbind(big_blending_frame)
+
+    se = H2OStackedEnsembleEstimator(
+        base_models=gs1.model_ids,
+        max_runtime_secs=max_runtime_secs,
+        blending_frame=big_blending_frame)
+    se.train(data.x, data.y, data.train)
+
+    assert se.metalearner() is not None
+
+    # We should have SE with just one base model due to time constrains; intercept + base_model ==> 2
```

Comment:
@tomasfryda I'm confused by this comment -- how does the metalearner time constraint impact the number of base models included in the Stacked Ensemble?

Reply:
The behaviour I implemented is as follows:

When we have a big blending frame, we  need to get predictions on that. If we run out of time during the predicting, we skip the rest of the base models. In the test, the situation is really just one extreme. 

I also considered another approach - if we ran out of time just fail. But from what I can tell about other algos they tend not to fail, rather they just "don't converge", e.g., DRF builds just the amount of trees it had time to etc.

I am open to suggestions how to handle this differently.

Reply:
Gotcha, I understand the `shouldStop` in `StackedEnsemble` now.
I'm in favor of failing in that case for a simple reason: other algos do not fail if the training was started. As you say, they just "don't converge".
But in your case here, the metalearner training didn't even start, so you'll start a metalearner training with only a random subset of the models, and  besides, as it's timing out, you'll train it for a couple of seconds max: no point, really, better skip it completely.

Reply:
Ok, I modified it back to just fail if there is not enough time. In R, I am throwing an error when I can't retrieve the metalearner saying it might be due to time out. In python, I don't do anything. The reason is that R uses functional interface and returns an object - something that we didn't build (as opposed to python's object oriented client interface).

I don't like that the two implementations behave differently. Should I also raise an exception from python? It didn't seem to me like something that other algos do.

Reply:
I see your concern now: other algos never (never?) fail on timeout, and starting the metalearner, even for 1s, avoided that as well.
I think raising an error on both R and Py clients make sense if metalearner is not available, you could even precise in your message that there was not enough time to start training the metalearner to inform user why this is failing instead of just "timing out".